vsphere/data_source_vsphere_datacenter_test.go

```diff
@@ -0,0 +1,100 @@
+package vsphere
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+var testAccDataSourceVSphereDatacenterExpectedRegexp = regexp.MustCompile("^datacenter-")
+
+func TestAccDataSourceVSphereDatacenter(t *testing.T) {
+	var tp *testing.T
+	testAccDataSourceVSphereDatacenterCases := []struct {
+		name     string
+		testCase resource.TestCase
+	}{
+		{
+			"basic",
+			resource.TestCase{
+				PreCheck: func() {
+					testAccPreCheck(tp)
+					testAccDataSourceVSphereDatacenterPreCheck(tp)
+				},
+				Providers: testAccProviders,
+				Steps: []resource.TestStep{
+					{
+						Config: testAccDataSourceVSphereDatacenterConfig(),
+						Check: resource.ComposeTestCheckFunc(
+							resource.TestMatchResourceAttr(
+								"data.vsphere_datacenter.dc",
+								"id",
+								testAccDataSourceVSphereDatacenterExpectedRegexp,
+							),
+							resource.TestMatchResourceAttr(
+								"data.vsphere_datacenter.dc",
+								"datacenter_id",
+								testAccDataSourceVSphereDatacenterExpectedRegexp,
+							),
+						),
+					},
+				},
+			},
+		},
+		{
+			"default",
+			resource.TestCase{
+				PreCheck: func() {
+					testAccPreCheck(tp)
+					testAccDataSourceVSphereDatacenterPreCheck(tp)
+				},
+				Providers: testAccProviders,
+				Steps: []resource.TestStep{
+					{
+						Config: testAccDataSourceVSphereDatacenterConfigDefault,
+						Check: resource.ComposeTestCheckFunc(
+							resource.TestMatchResourceAttr(
+								"data.vsphere_datacenter.dc",
+								"id",
+								testAccDataSourceVSphereDatacenterExpectedRegexp,
+							),
+							resource.TestMatchResourceAttr(
+								"data.vsphere_datacenter.dc",
+								"datacenter_id",
+								testAccDataSourceVSphereDatacenterExpectedRegexp,
+							),
+						),
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testAccDataSourceVSphereDatacenterCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tp = t
+			resource.Test(t, tc.testCase)
+		})
+	}
+}
+
+func testAccDataSourceVSphereDatacenterPreCheck(t *testing.T) {
+	if os.Getenv("VSPHERE_DATACENTER") == "" {
+		t.Skip("set VSPHERE_DATACENTER to run vsphere_datacenter acceptance tests")
+	}
+}
+
+func testAccDataSourceVSphereDatacenterConfig() string {
+	return fmt.Sprintf(`
+data "vsphere_datacenter" "dc" {
+  name = "%s"
+}
+`, os.Getenv("VSPHERE_DATACENTER"))
```

Comment:
Hopefully this isn't too nitpicky but I think that `os.Getenv("VSPHERE_DATACENTER"))` should be passed into the function instead of being called here. It'd be `func testAccDataSourceVSpohereDatacenterConfig(datacenter string) string {...}`

Reply:
I'm kind of on the fence with this one - one one hand I can see the merit of the full parameterization, allowing more flexibility of the generated config, on the other hand the only reason config is being generated in this case is to ensure that `VSPHERE_DATACENTER` can be passed in, otherwise it'd be a constant config like the default test, but with a static value for `name`. What I'd hate to see is this be a pattern where test cases end up scaling out and something like this has to be pasted in every test:

```
Config: testAccDataSourceVSphereDatacenterConfig(os.Getenv("VSPHERE_DATACENTER")),
```

Especially if the vast majority of the configs would take the parameter like that, it would start to seem redundant. Thoughts?

Reply:
I can stand behind that! I was also on the fence when I sent it away but you make valid points so I'm cool with keeping it this way!